File: samples/XMP/Laplace/xmp_laplace.c
#include <stdio.h> 
#include <stdlib.h> 
#include <math.h>
#define N1 64
#define N2 64
double u[N2][N1], uu[N2][N1];

#pragma xmp nodes p[*][4]
#pragma xmp template t[N2][N1]
#pragma xmp distribute t[block][block] onto p
#pragma xmp align u[j][i] with t[j][i]
#pragma xmp align uu[j][i] with t[j][i]
#pragma xmp shadow uu[1:1][1:1]

int main(int argc, char **argv)
{
  int i, j, k, niter = 100;
  double value = 0.0;

#pragma xmp loop (j,i) on t[j][i]
  for(j = 0; j < N2; j++){
    for(i = 0; i < N1; i++){
      u[j][i] = 0.0;
      uu[j][i] = 0.0;
    }
  }

#pragma xmp loop (j,i) on t[j][i]
  for(j = 1; j < N2-1; j++)
    for(i = 1; i < N1-1; i++)
      u[j][i] = sin((double)i/N1*M_PI) + cos((double)j/N2*M_PI);
  
  for(k = 0; k < niter; k++){

#pragma xmp loop (j,i) on t[j][i]
    for(j = 1; j < N2-1; j++)
      for(i = 1; i < N1-1; i++)
	uu[j][i] = u[j][i];
    
#pragma xmp reflect (uu)

#pragma xmp loop (j,i) on t[j][i]
    for(j = 1; j < N2-1; j++)
      for(i = 1; i < N1-1; i++)
	u[j][i] = (uu[j-1][i] + uu[j+1][i] + uu[j][i-1] + uu[j][i+1])/4.0;
  }
  
#pragma xmp loop (j,i) on t[j][i] reduction(+:value)
  for(j = 1; j < N2-1; j++)
    for(i = 1; i < N1-1; i++)
      value += fabs(uu[j][i] - u[j][i]);

#pragma xmp task on p[0][0]
  printf("Verification = %20.16f\n", value);

  return 0;
}

File: samples/XMP/Laplace/xmp_laplace.f90
program laplace
  integer,parameter :: N1=64, N2=64
  real(8),parameter :: PI = 3.141592653589793238463

  integer :: niter=100
  real(8) :: u(N1,N2),uu(N1,N2)
  real(8) :: value = 0.0

!$xmp nodes p(4,*)
!$xmp template t(N1,N2)
!$xmp distribute t(block, block) onto p
!$xmp align u(i,j) with t(i, j)
!$xmp align uu(i,j) with t(i, j)
!$xmp shadow uu(1:1,1:1)

!$xmp loop (i,j) on t(i,j)
  do j=1,N2
     do i=1,N1
        u(i,j)=0.0
        uu(i,j)=0.0
     end do
  end do

!$xmp loop (i,j) on t(i,j)
  do j=2,N2-1
     do i=2,N1-1
        u(i,j)=sin(dble(i-1)/N1*PI)+cos(dble(j-1)/N2*PI)
     end do
  end do

  do k=1,niter

!$xmp loop (i,j) on t(i,j)
     do j=2,N2-1
        do i=2,N1-1
           uu(i,j)=u(i,j)
        end do
     end do

!$xmp reflect (uu)

!$xmp loop (i,j) on t(i,j)
     do j=2,N2-1
        do i=2,N1-1
           u(i,j)=(uu(i-1,j) + uu(i+1,j) + uu(i,j-1) + uu(i,j+1))/4.0
        end do
     end do

  enddo

!$xmp loop (i,j) on t(i,j) reduction(+:value)
  do j=2,N2-1
     do i=2,N1-1
        value = value + dabs(uu(i,j) -u(i,j))
     end do
  end do

!$xmp task on p(1,1)
  print *, 'Verification =', value
!$xmp end task
  
end program laplace
